src/Compilers/Core/Portable/Diagnostic/Diagnostic_SimpleDiagnostic.cs

```diff
@@ -130,11 +130,15 @@ public override bool Equals(object obj)
 
             public override int GetHashCode()
             {
-                return Hash.Combine(_descriptor,
-                        Hash.Combine(_messageArgs.GetHashCode(),
-                         Hash.Combine(_location.GetHashCode(),
-                          Hash.Combine(_severity.GetHashCode(), _warningLevel)
-                        )));
+                int hash = 0;
+
+                hash = Hash.Combine(_location.GetHashCode(), 0);
+                hash = Hash.Combine((int)_severity, hash);
```

Comment:
I have not looked at that. But relying on the caller to know that location cannot be null does not seem like a good idea if the first time an exception is thrown is when GetHashCode is called